### ISSUE TITLE
[Bug] Engine.generate(return_logprob=True) — host RAM blowup and throughput collapse from unbucketed logprob path

### ISSUE THREAD

**Issue by MokusMokun** [sgl-project/sglang-jax#1176]
## Symptom

`Engine.generate(..., return_logprob=True, logprob_start_len=0)` causes the scheduler subprocess (`sglang::scheduler_NP0`) host RSS to grow until the pod is OOM-killed, and throughput collapses to a small fraction of the no-logprob baseline. Reproduced on TPU v6e 4×4 with Qwen3-30B-A3B (vocab=151,936), driving a long-running, heterogeneous-shape eval workload (MMLU-Pro PPL-style: ~13.5k prompts × 14 categories).

Observed end-to-end with a local fix in place: throughput recovers ~19× (from ~0.025 q/s pre-fix to ~0.47 q/s post-fix), because the scheduler main thread was previously blocked inside `jax/_src/compilation_cache.py::compress_executable` / `_cache_write` between most batches.

## Root cause — unbucketed logprob path with two compounded symptoms

The `return_logprob=True` extend path in `LogitsProcessor.__call__` bypasses the bucketed-shape design used by the main prefill path. The main path pre-pads `input_ids` / `positions` / `out_cache_loc` to `total_token_size` so the JIT cache key stays stable across batches; the logprob path does not, and additionally returns full vocab rows to host. Two symptoms result.

### Symptom 1 — full-vocab `input_token_logprobs` returned to host (true leak)

`meta_info["input_token_logprobs"]` is documented as a list of `(scalar_logprob, token_id, text)` per input token. In practice each entry carries `(vocab_logprobs[1, vocab_size], token_id, text)` — the entire vocab row. For vocab=151,936 × bf16 that is **~297 KB per token**, ~178 MB per ~600-tok prompt, ~36 GB per 200-prompt `generate()` call. The host transfer at `scheduler_output_processor_mixin.py:86-88` does `jax.device_get(logits_output.input_token_logprobs).astype(float)` on the full-vocab tensor; nothing on the path narrows it to the realized token before the host transfer.

### Symptom 2 — JIT cache key instability causes recompile churn (not a leak, but unbounded host growth)

This is **not** a true leak — JAX's live-executable cache has no LRU eviction, so every distinct cache key retains its compiled artifact (~800 MB of Pallas IR + HLO + jaxpr) on host indefinitely. With a churning key, host RSS grows linearly (~1.3 GB/min in our workload) and only plateaus once the key space is exhausted, which never happens on heterogeneous workloads. Two independent shape-instability sources stack:

1. **Aux-data instability.** `LogitsMetadata.tree_flatten` places `extend_seq_lens_cpu` / `extend_logprob_start_lens_cpu` / `extend_logprob_pruned_lens_cpu` (Python `list[int]`) in the pytree aux dict. JAX compares aux by value, so any per-batch variation in these lists is a new cache key.
2. **Traced-shape instability.** Even with aux fixed, `LogitsProcessor.__call__` builds `pruned_states = jnp.concat([hidden_states[pt+start:pt+end] for ...])`, making the traced intermediate length equal to `sum(extend_len - start_len)` — which varies per batch. So the abstract shape entering `_get_logits` is also a per-batch cache key.

Together these drive a sustained ~4% recompile rate on prefill batches. `py-spy dump` caught the scheduler main thread inside `compress_executable` mid-run — the cost shows up in throughput as well as RSS, because the main loop is blocked compiling rather than scheduling the next batch.

## Reproduction

check https://github.com/sgl-project/sglang-jax/issues/1176#issuecomment-4507231466

## Fix

PR: https://github.com/sgl-project/sglang-jax/pull/1177

Results: https://github.com/sgl-project/sglang-jax/issues/1176#issuecomment-4508284599

**Comment by MokusMokun** [comment]
Fix PR: #1177 (draft, pending CI label).

**Comment by MokusMokun** [comment]
Reproduced on `main` (`82f0bd5`) on a TPU v6e 4×4 with **Qwen3-30B-A3B** (vocab=151,936), using a minimal synthetic workload (200-prompt batch, heterogeneous lengths ~60–600 words, `return_logprob=True`, `logprob_start_len=0`). All three claims in the issue body match the measurements.

### S1 — full-vocab tensor returned to host

Single-prompt probe immediately after warmup:

```
[S1 probe] type=ndarray shape=(1, 151936) is_scalar=False
[S1 probe] BUG REPRODUCED: input_token_logprobs[i][0] is a tensor of shape (1, 151936)
```

`meta_info["input_token_logprobs"][i][0]` is the full vocab row, not the scalar promised by the docstring.

### S2 — JIT cache key churn + linear host-RAM growth

Two consecutive rounds of the same 50-prompt workload (different prompt-length seeds):

```
[S2 baseline]   jit_cache_entries=680   scheduler RSS=  9530 MB
[S2 round 0]    jit_cache_entries=759  (+79)        RSS= 33738 MB  (+24,208)   thr=0.062 q/s
[S2 round 1]    jit_cache_entries=829 (+149)        RSS= 50972 MB  (+41,442)   thr=0.066 q/s
```

`jit_cache_entries` is the file count under `JAX_COMPILATION_CACHE_DIR` — each cache miss writes one entry. ~75 new entries per 50 prompts means **~1.5 new compiled artifacts per prompt**, and the scheduler `Prefill batch` log reports `#cache_miss: 2` on essentially every batch — consistent with the two stacked instabilities (aux-dict and traced-shape) described in the root-cause section.

Quantitative match against the issue body:

| Claim in issue | Measured |
|---|---|
| `[1, vocab_size]` full-vocab return | `(1, 151936)` ✓ |
| Scheduler RSS grows linearly at ~1.3 GB/min | `(50972 − 9530) MB / (807 + 758) s` ≈ **1.59 GB/min** ✓ |
| Throughput collapses to a small fraction of baseline | 0.06 q/s vs an expected several-q/s prefill rate ✓ |
| Sustained recompiles on prefill batches | `#cache_miss: 2` on essentially every prefill batch ✓ |

If the workload were extended to ~14 categories × 200 prompts (the original repro shape), the +41 GB / 2 rounds trajectory crosses the typical 700 GB host ceiling within hours — matching the OOM behavior in the symptom section.

### Repro script

Pure-synthetic, public-API only — no datasets. Posting here as a self-contained regression check that can be folded into the fix PR.

```python
"""Repro for sgl-project/sglang-jax#1176.

Triggers and measures both symptoms of the unbucketed-logprob path on a
4-host TPU v6e 4×4 pod. Pure synthetic prompts, public Engine API only.

  NODE_RANK=$i DIST_INIT_ADDR=$RANK0_IP:29500 \
  N_PROMPTS=50 N_ROUNDS=2 JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache \
  python -u repro_logprob_leak.py /path/to/Qwen3-30B-A3B
"""

import gc
import os
import random
import sys
import time
from pathlib import Path

import psutil

from sgl_jax.srt.entrypoints.engine import Engine


def _scheduler_rss_mb(pid: int) -> float:
    return psutil.Process(pid).memory_info().rss / 1024 / 1024


def _jit_cache_entries(cache_dir: str) -> int:
    p = Path(cache_dir)
    if not p.exists():
        return 0
    return sum(1 for _ in p.rglob("*") if _.is_file())


def _build_prompts(n: int, seed: int = 0) -> list[str]:
    rng = random.Random(seed)
    return [("hello " * rng.randint(60, 600)).strip() for _ in range(n)]


def main():
    model_path = sys.argv[1]
    node_rank = int(os.environ.get("NODE_RANK", "0"))
    dist_init_addr = os.environ.get("DIST_INIT_ADDR")
    n_prompts = int(os.environ.get("N_PROMPTS", "200"))
    n_rounds = int(os.environ.get("N_ROUNDS", "3"))
    jit_cache_dir = os.environ.get("JAX_COMPILATION_CACHE_DIR", "/tmp/jit_cache")
    if dist_init_addr is None:
        raise SystemExit("DIST_INIT_ADDR env var is required")

    llm = Engine(
        model_path=model_path,
        trust_remote_code=True,
        tp_size=16,
        nnodes=4,
        node_rank=node_rank,
        dist_init_addr=dist_init_addr,
        dp_size=1,
        device="tpu",
        dtype="bfloat16",
        mem_fraction_static=0.8,
        chunked_prefill_size=512,
        page_size=256,
        max_running_requests=64,
        disable_radix_cache=True,
        disable_overlap_schedule=True,
        precompile_bs_paddings=[1, 16, 64],
        precompile_token_paddings=[16, 128, 512],
        context_length=4096,
        skip_server_warmup=True,
        log_level="info",
    )

    if node_rank != 0:
        time.sleep(36_000)
        return

    scheduler_pid = None
    for p in psutil.process_iter(attrs=["pid", "name"]):
        if "scheduler" in (p.info["name"] or ""):
            scheduler_pid = p.info["pid"]
            break

    print(f"[repro] scheduler_pid={scheduler_pid}", flush=True)
    print(f"[repro] initial jit_cache_entries={_jit_cache_entries(jit_cache_dir)}", flush=True)
    print(f"[repro] initial scheduler RSS={_scheduler_rss_mb(scheduler_pid):.1f} MB", flush=True)

    # S1: a single small batch shows the full-vocab row on input_token_logprobs[i][0]
    probe = llm.generate(
        ["hello " * 200],
        {"max_new_tokens": 1, "temperature": 0, "top_p": 1},
        return_logprob=True,
        logprob_start_len=0,
    )
    lp_val = probe[0]["meta_info"]["input_token_logprobs"][1][0]
    shape = getattr(lp_val, "shape", None)
    is_scalar = isinstance(lp_val, (float, int)) or (shape == () or shape is None)
    print(f"[S1 probe] type={type(lp_val).__name__} shape={shape} is_scalar={is_scalar}", flush=True)
    if not is_scalar:
        print(f"[S1 probe] BUG REPRODUCED: input_token_logprobs[i][0] is shape {shape}", flush=True)

    # S2: heterogeneous prompts over multiple rounds — watch cache entries and RSS grow
    base_live = _jit_cache_entries(jit_cache_dir)
    base_rss = _scheduler_rss_mb(scheduler_pid)
    print(f"[S2] baseline jit_cache_entries={base_live}, scheduler RSS={base_rss:.1f} MB", flush=True)

    for r in range(n_rounds):
        prompts = _build_prompts(n_prompts, seed=r)
        t0 = time.time()
        outputs = llm.generate(
            prompts,
            {"max_new_tokens": 1, "temperature": 0, "top_p": 1},
            return_logprob=True,
            logprob_start_len=0,
        )
        dt = time.time() - t0
        live_after = _jit_cache_entries(jit_cache_dir)
        rss_after = _scheduler_rss_mb(scheduler_pid)
        print(
            f"[S2 round {r}] n={len(prompts)} took={dt:.1f}s thr={len(prompts)/dt:.3f} q/s "
            f"jit_cache_entries={live_after} (+{live_after - base_live}) "
            f"scheduler RSS={rss_after:.1f} MB (+{rss_after - base_rss:.1f})",
            flush=True,
        )
        del outputs
        gc.collect()

    print("[repro] done", flush=True)
    llm.shutdown()


if __name__ == "__main__":
    main()
```

**Comment by MokusMokun** [comment]
Verified PR #1177 (`1712b0b`) on the same TPU v6e 4×4 + Qwen3-30B-A3B setup with the unchanged repro script. JIT cache was cleared first so the count starts from a clean baseline.

| | Pre-fix (`82f0bd5`) | Post-fix (`1712b0b`) |
|---|---|---|
| `input_token_logprobs[i][0]` | `ndarray (1, 151936)` | `float`, `is_scalar=True` |
| S2 round 0 / 1 throughput | 0.062 / 0.066 q/s | **29.87 / 29.80 q/s** (~450×) |
| New JIT cache entries (round 0 / 1) | +79 / +149 | **+0 / +0** |
| Scheduler RSS delta (round 0 / 1) | +24,208 MB / +41,442 MB | **+0.8 MB / +0.9 MB** |
| `#cache_miss` per prefill batch | 2 on essentially every batch | 0 after warmup |

All four acceptance criteria pass. LGTM from a repro standpoint.
